src/Microsoft.SqlTools.ServiceLayer/Connection/ListDatabaseRequestHandler.cs

```diff
@@ -57,43 +61,66 @@ abstract class ListDatabaseRequestHandler<T> : IListDatabaseRequestHandler
-        public abstract string QueryText { get; }
-
         public ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo)
         {
             ConnectionDetails connectionDetails = connectionInfo.ConnectionDetails.Clone();
+            var serverConnection = ConnectionService.OpenServerConnection(connectionInfo);
+
+            SMO.Server SQLServer = new SMO.Server(serverConnection);
+
 
-            // Connect to master
-            connectionDetails.DatabaseName = "master";
-            using (var connection = connectionFactory.CreateSqlConnection(ConnectionService.BuildConnectionString(connectionDetails), connectionDetails.AzureAccountToken))
+            List<T> result = new List<T>();
+            ListDatabasesResponse response = new ListDatabasesResponse();
+
+
+            foreach (SMO.Database db in SQLServer.Databases)
             {
-                connection.Open();
-                ListDatabasesResponse response = new ListDatabasesResponse();
-                using (DbCommand command = connection.CreateCommand())
+                if (db.Name != serverConnection.CurrentDatabase)
                 {
-                    command.CommandText = this.QueryText;
-                    command.CommandTimeout = 15;
-                    command.CommandType = CommandType.Text;
-                    using (var reader = command.ExecuteReader())
+                    if(serverConnection.IsContainedAuthentication){ //Check if the user is a contained user. If yes, then don't add the database.
+                        continue;
+                    }
+
+                    ServerConnection dbServerConnection;
+                    dbServerConnection = serverConnection.GetDatabaseConnection(db.Name);
+
+                    SqlExecutionModes originalMode = serverConnection.SqlExecutionModes;
+                    dbServerConnection.SqlExecutionModes = SqlExecutionModes.ExecuteSql;
+
+                    try
+                    {
+                        dbServerConnection.ExecuteNonQuery(String.Format(CultureInfo.InvariantCulture ,"use [{0}]", Util.EscapeString(db.Name , ']')));
+                        result.Add(CreateItem(db));
+                    }
+                    catch (Exception exc)
                     {
-                        List<T> results = new List<T>();
-                        while (reader.Read())
+                        SqlException sqlExc = exc.InnerException as SqlException;
+                        if (sqlExc != null
+                            && (sqlExc.Number == 916 //The server principal is not able to access the database under the current security context.
+                                || sqlExc.Number == 911) //Database does not exist. Make sure that the name is entered correctly.
+                            )
                         {
-                            results.Add(this.CreateItem(reader));
+                            Console.WriteLine(exc.Message);
```

Comment:
Don't write to console, that isn't very helpful here. Use the output logger and make sure to add context so we know where the error is coming from